subsys/net/l2/ethernet/ethernet.c

```diff
@@ -735,14 +735,10 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 arp_error:
 	if (IS_ENABLED(CONFIG_NET_ARP) && ptype == htons(NET_ETH_PTYPE_ARP)) {
 		/* Original packet was added to ARP's pending Q, so, to avoid it
-		 * being freed, take a reference, the reference is dropped when we
-		 * clear the pending Q in ARP and then it will be freed by net_if.
+		 * being freed, take a reference, the reference is dropped by net_if,
+		 * then it will be freed when clearing the ARP pending Q as timeout.
 		 */
 		net_pkt_ref(orig_pkt);
-		if (net_arp_clear_pending(
```

Comment:
Thanks for your comment, let me re-organize my explanations (part from jukkar):
In ethernet_send(), if ARP packets send failed, original code tried to clean up the entry (i.e, clear the entry address) from ARP pending queue, but the entry wasn't removed from the arp_pending_entries, until the ARP request was timeout.
Within this timeout period, this ARP pending entry can't be found due to the entry address mismatch.
If we add ‘entry removed from arp_pending_entries and append to arp_free_entries' here, it should also add arp_mutex protection like other functions in arp.c. But add arp_mutex protection will have side effect. For example, in the net_arp_update(), it will take the arp_mutex first, then may call the ethernet_send(), so add other arp_mutex protection in ethernet_send() is not a good choice and may lead to dead lock.
The better way is just removing the net_arp_clear_pending() in ethernet_send(), and let the ARP request timeout to do the entry cleanup, removed and append to arp_free_entries at the same time.

Reply:
Thank you for the detailed explanation.

> In ethernet_send(), if ARP packets send failed, original code tried to clean up the entry (i.e, clear the entry address) from ARP pending queue, but the entry wasn't removed from the arp_pending_entries, until the ARP request was timeout.
Within this timeout period, this ARP pending entry can't be found due to the entry address mismatch.
If we add ‘entry removed from arp_pending_entries and append to arp_free_entries' here, it should also add arp_mutex protection like other functions in arp.c

That's correct, the original code we have right now is definitely flawed in that regard.

> But add arp_mutex protection will have side effect. For example, in the net_arp_update(), it will take the arp_mutex first, then may call the ethernet_send(), so add other arp_mutex protection in ethernet_send() is not a good choice and may lead to dead lock.

So I'm not really sure if the actual concern about the deadlock is a valid one in this particular case. In order for deadlock to take place, we'd need to have two mutexes, locked in an alternate way from two different threads. But in the scenario you describe, it's more about locking the same mutex for the second time from the same thread. And this isn't really a concern, as the mutex implementation is smart enough to identify that if the mutex lock was called from the thread that already hold the mutex, it should not block. I couldn't really identify any other mutex that could contribute to the possible deadlock scenario in this case.

Especially, that on the Ethernet TX path, we already have `net_arp_prepare()` being called from `ethernet_ll_prepare_on_ipv4()`, which locks the ARP mutex, so it shouldn't be really different if we add mutex protection to `net_arp_clear_pending()` and actually update the lists.

> The better way is just removing the net_arp_clear_pending() in ethernet_send(), and let the ARP request timeout to do the entry cleanup, removed and append to arp_free_entries at the same time.

My only concern with this approach is that with the timeout approach, we keep the original TX packet allocated until the ARP entry times out, which isn't great. I think it'd be better to fix `net_arp_clear_pending()` instead.

Reply:
Thanks for your advice. The better way I think I just keep ethernet_send() unchanged, to avoid introducing more side effect.
Actually in my case, issue is caused by arp_prepare() bug. About net_arp_clear_pending() called in ethernet_send(), it's just my personal concern when reading code.

Reply:
please help review again